# HG changeset patch
# User Philip Nienhuis <prnienhuis@users.sf.net>
# Date 1439155395 -7200
#      Sun Aug 09 23:23:15 2015 +0200
# Branch stable
# Node ID 4c4d8fe5583a008d4436afd8666eb8e2816c469a
# Parent  97edbce9b975166eb7362632f7925284db149b0c
strread.m: properly preprocess all forms of string format specifier (bug #45712)

diff --git a/scripts/io/strread.m b/scripts/io/strread.m
--- a/scripts/io/strread.m
+++ b/scripts/io/strread.m
@@ -383,17 +383,17 @@ function varargout = strread (str, forma
     end
     str = cellslices (str, [1, cstop + c2len], [cstart - 1, len]);
     str = [str{:}];
   endif
 
   if (! isempty (white_spaces))
     ## For numeric fields, whitespace is always a delimiter, but not for text
     ## fields
-    if (isempty (strfind (format, "%s")))
+    if (isempty (regexp (format, '%\*?\d*s')))
       ## Add whitespace to delimiter set
       delimiter_str = unique ([white_spaces delimiter_str]);
     else
       ## Remove any delimiter chars from white_spaces list
       white_spaces = setdiff (white_spaces, delimiter_str);
     endif
   endif
   if (isempty (delimiter_str))
@@ -1081,8 +1081,14 @@ endfunction
 
 ## Illegal format specifiers
 %!test
 %!error <no valid format conversion specifiers> strread ("1.0", "%z");
 
 ## Test for false positives in check for non-supported format specifiers
 %!test
 %! assert (strread ("Total: 32.5 % (of cm values)","Total: %f % (of cm values)"), 32.5, 1e-5);
+
+## Test various forms of string format specifiers (bug #45712)
+%!test
+%! str = "14 :1 z:2 z:3 z:5 z:11";
+%! [a, b, c, d] = strread (str, "%f %s %*s %3s %*3s %f", "delimiter", ":");
+%! assert ({a, b, c, d}, {14, {"1 z"}, {"3 z"}, 11});
